nexus/src/sagas.rs

```diff
@@ -308,3 +335,332 @@ async fn sic_instance_ensure(
+    Ok(())
+}
+
+async fn sdc_alloc_regions(
+    sagactx: ActionContext<SagaDiskCreate>,
+) -> Result<Vec<(db::model::Dataset, db::model::Region)>, ActionError> {
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params();
+    let disk_id = sagactx.lookup::<Uuid>("disk_id")?;
+    // Ensure the disk is backed by appropriate regions.
+    //
+    // This allocates regions in the database, but the disk state is still
+    // "creating" - the respective Crucible Agents must be instructed to
+    // allocate the necessary regions before we can mark the disk as "ready to
+    // be used".
+    let datasets_and_regions = osagactx
+        .datastore()
+        .region_allocate(disk_id, &params.create_params)
+        .await
+        .map_err(ActionError::action_failed)?;
+    Ok(datasets_and_regions)
+}
+
+async fn sdc_alloc_regions_undo(
+    sagactx: ActionContext<SagaDiskCreate>,
+) -> Result<(), anyhow::Error> {
+    let osagactx = sagactx.user_data();
+
+    let disk_id = sagactx.lookup::<Uuid>("disk_id")?;
+    osagactx.datastore().regions_hard_delete(disk_id).await?;
+    Ok(())
+}
+
+async fn ensure_region_in_dataset(
+    log: &Logger,
+    dataset: &db::model::Dataset,
+    region: &db::model::Region,
+) -> Result<crucible_agent_client::types::Region, Error> {
+    let url = format!("http://{}", dataset.address());
+    let client = CrucibleAgentClient::new(&url);
+
+    let region_request = CreateRegion {
+        block_size: region.block_size(),
+        extent_count: region.extent_count(),
+        extent_size: region.extent_size(),
+        // TODO: Can we avoid casting from UUID to string?
+        // NOTE: This'll require updating the crucible agent client.
+        id: RegionId(region.id().to_string()),
+        volume_id: region.disk_id().to_string(),
+    };
+
+    let create_region = || async {
+        let region = client
+            .region_create(&region_request)
+            .await
+            .map_err(|e| BackoffError::Permanent(e))?;
+        match region.state {
+            RegionState::Requested => Err(BackoffError::Transient(anyhow!(
+                "Region creation in progress"
+            ))),
+            RegionState::Created => Ok(region),
+            _ => Err(BackoffError::Permanent(anyhow!(
+                "Failed to create region, unexpected state: {:?}",
+                region.state
+            ))),
+        }
+    };
+
+    let log_create_failure = |_, delay| {
+        warn!(
+            log,
+            "Region requested, not yet created. Retrying in {:?}", delay
+        );
+    };
+
+    let region = backoff::retry_notify(
+        backoff::internal_service_policy(),
+        create_region,
+        log_create_failure,
+    )
+    .await?;
+
+    Ok(region)
+}
+
+async fn sdc_regions_ensure(
+    sagactx: ActionContext<SagaDiskCreate>,
+) -> Result<(), ActionError> {
+    let log = sagactx.user_data().log();
+    let datasets_and_regions = sagactx
```

Comment:
I don't know, because at the moment I believe Crucible only supports a single RegionSet. This means that the maximum here is currently "3", and I have no idea what the lower/upper bounds are on RegionSets.

If it's okay with you, I'm going to include a link to https://github.com/oxidecomputer/omicron/issues/613 as a TODO and punt until we get that integration.